### PR TITLE
Refactor tests, remove fixtures

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

Follow-up to #4944 

*Description of changes:*

Previously, it was very hard to use the unit test logic outside of AutoGluon-Tabular's unit test suite due to the usage of fixtures. I've now removed the usage of fixtures and moved the core logic that tests models into `autogluon.tabular.testing`, which is publicly accessible and part of the `src` folder instead of the `tests` folder.

Now users can easily access and use these utilities to check if a custom model works in a wide-range of scenarios and edge-cases.

For example:

```python
from autogluon.tabular.testing import FitHelper
from tabrepo.benchmark.models import RealMLPModel

FitHelper.verify_model(model_cls=RealMLPModel, model_hyperparameters={})
```

Additionally, I cleaned up the `test_tabular.py` file, removing all redundant / overly verbose code. The functionality of the tests have remained identical.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tabular/src/autogluon/tabular/testing/fit_helper.py [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+        if not os.path.exists(directory_prefix):
+            os.mkdir(directory_prefix)
+        directory = directory_prefix + name + "/"
+        train_file_path = directory + train_file
+        test_file_path = directory + test_file
+        if (not os.path.exists(train_file_path)) or (not os.path.exists(test_file_path)):
+            # fetch files from s3:
+            print("%s data not found locally, so fetching from %s" % (name, url))
+            zip_name = download(url, directory_prefix)
+            unzip(zip_name, directory_prefix)
```
Comment (shchur) [inline tabular/src/autogluon/tabular/testing/fit_helper.py:122]: Are files assumed to always be zipped?

File: tabular/src/autogluon/tabular/testing/fit_helper.py [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+        compile: bool = False,
+        compiler_configs: dict | None = None,
+        allowed_dataset_features: list[str] | None = None,
+        expected_stacked_overfitting_at_test: bool | None = None,
+        expected_stacked_overfitting_at_val: bool | None = None,
+        scikit_api: bool = False,
+        use_test_data: bool = False,
+        use_test_for_val: bool = False,
+        raise_on_model_failure: bool | None = None,
+        deepcopy_fit_args: bool = True,
```
Comment (shchur) [inline tabular/src/autogluon/tabular/testing/fit_helper.py:155]: Is there a reason not to always do the deep copy?

**Review by shchur — APPROVED** [review]
A few minor comments/questions, but otherwise LGTM

File: tabular/src/autogluon/tabular/testing/fit_helper.py (hunk 2 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/testing/fit_helper.py:155]: Yes: For callback objects, we want to check the state of the callback after fitting, so avoiding deepcopy lets us view the callback object after updates to ensure the logic is working correctly.

File: tabular/src/autogluon/tabular/testing/fit_helper.py (hunk 1 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/testing/fit_helper.py:122]: This is based on the old logic from 2019, simply copy pasted out of tests. It should ideally be improved, but this is a refactor for another time.

File: tabular/src/autogluon/tabular/testing/fit_helper.py [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+    def load_data(
+        directory_prefix: str,
+        train_file: str,
+        test_file: str,
+        name: str,
+        url: str | None = None,
+    ):
+        if not os.path.exists(directory_prefix):
+            os.mkdir(directory_prefix)
+        directory = directory_prefix + name + "/"
```
[the remark this answers is not in the document]
Comment (Innixma) [reply tabular/src/autogluon/tabular/testing/fit_helper.py:115]: I suppose it does, this is old logic from 2019. It works on Windows since the tests pass on Windows, but can be improved in future refactors.

File: tabular/src/autogluon/tabular/testing/fit_helper.py [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+        toy_quantile=generate_toy_quantile_dataset,
+        toy_binary_10=generate_toy_binary_10_dataset,
+        toy_multiclass_10=generate_toy_multiclass_10_dataset,
+        toy_regression_10=generate_toy_regression_10_dataset,
+        toy_quantile_10=generate_toy_quantile_10_dataset,
+        toy_multiclass_30=generate_toy_multiclass_30_dataset,
+    )
+
+    @staticmethod
+    def load_dataset_toy(name: str) -> tuple[pd.DataFrame, pd.DataFrame, dict]:
```
[the remark this answers is not in the document]
Comment (Innixma) [reply tabular/src/autogluon/tabular/testing/fit_helper.py:83]: thanks! It is unnecessary, removed and merged into load_dataset

File: tabular/src/autogluon/tabular/testing/fit_helper.py [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+        return train_data, test_data, dataset_info
+
+    @staticmethod
+    def load_data(
+        directory_prefix: str,
+        train_file: str,
+        test_file: str,
+        name: str,
+        url: str | None = None,
+    ):
```
[the remark this answers is not in the document]
Comment (Innixma) [reply tabular/src/autogluon/tabular/testing/fit_helper.py:112]: added

File: tabular/src/autogluon/tabular/testing/model_fit_helper.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+from autogluon.core.data.label_cleaner import LabelCleaner
+from autogluon.core.models import AbstractModel, BaggedEnsembleModel
+from autogluon.core.utils import generate_train_test_split, infer_problem_type
+from autogluon.features.generators import AbstractFeatureGenerator, AutoMLPipelineFeatureGenerator
+
+from autogluon.tabular.testing.fit_helper import FitHelper
+
+
+# Helper functions for training models outside of predictors
+class ModelFitHelper:
```
[the remark this answers is not in the document]
Comment (Innixma) [reply tabular/src/autogluon/tabular/testing/model_fit_helper.py:17]: good point, added docstrings to clarify

File: tabular/src/autogluon/tabular/testing/fit_helper.py [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+        )
+
+        return train_data, test_data, dataset_info
+
+    @staticmethod
+    def load_data(
+        directory_prefix: str,
+        train_file: str,
+        test_file: str,
+        name: str,
```
[the remark this answers is not in the document]
Comment (Innixma) [reply tabular/src/autogluon/tabular/testing/fit_helper.py:110]: Added docstring and TODO calling for eventual refactor
